### PR TITLE
Flyin - Enhance code server installation based on system architecture

### PR DESCRIPTION
**Pull request by jasonlai1218**
## Why are the changes needed?
These changes are necessary to increase the flexibility of the get_code_server_info function in selecting the appropriate code server based on different system architectures, including commonly used AMD64/ARM64 architectures.

## What changes were proposed in this pull request?
- Modify `flytekit-flyin/flytekitplugins/flyin/__init__.py`:
  - Replace `DEFAULT_CODE_SERVER_DIR_NAME` with `DEFAULT_CODE_SERVER_DIR_NAMES`
  - Replace `DEFAULT_CODE_SERVER_REMOTE_PATH` with `DEFAULT_CODE_SERVER_REMOTE_PATHS`
- Modify `flytekit-flyin/flytekitplugins/flyin/vscode_lib/config.py`:
  - Replace `code_server_dir_name` with `code_server_dir_names`
  - Replace `code_server_remote_path` with `code_server_remote_paths`
- Modify `flytekit-flyin/flytekitplugins/flyin/vscode_lib/constants.py`:
  - Replace `DEFAULT_CODE_SERVER_REMOTE_PATH` with `DEFAULT_CODE_SERVER_REMOTE_PATHS`
  - Replace `DEFAULT_CODE_SERVER_DIR_NAME` with `DEFAULT_CODE_SERVER_DIR_NAMES`
- Modify `flytekit-flyin/flytekitplugins/flyin/vscode_lib/decorator.py`:
  - Add `get_code_server_info` function
  - Replace `config.code_server_remote_path` with `code_server_remote_path`
  - Replace `config.code_server_dir_name` with `code_server_dir_name`
  

## How was this patch tested?
- Modify `test_flyin_plugin.py` in the `plugins/flytekit-flyin/tests` directory
- Add `DEFAULT_CODE_SERVER_DIR_NAMES` and `DEFAULT_CODE_SERVER_REMOTE_PATHS` to the `flytekit-flyin` plugin configuration
- Add `mock_code_server_info_dict` function in the `plugins/flytekit-flyin/tests/test_flyin_plugin.py`
- Add `test_arm_platform` function in the `plugins/flytekit-flyin/tests/test_flyin_plugin.py`
- Add `test_amd_platform` function in the `plugins/flytekit-flyin/tests/test_flyin_plugin.py`
- Add `test_platform_unsupported` function in the `plugins/flytekit-flyin/tests/test_flyin_plugin.py`

### Setup process
I checkout my feature branch and wrote a Dockerfile
```
# This Dockerfile is here to help with end-to-end testing
# From flytekit
# $ docker build -f Dockerfile.dev --build-arg PYTHON_VERSION=3.10 -t localhost:30000/flytekittest:someversion .
# $ docker push localhost:30000/flytekittest:someversion
# From your test user code
# $ pyflyte run --image localhost:30000/flytekittest:someversion

ARG PYTHON_VERSION
FROM python:${PYTHON_VERSION}-slim-buster

MAINTAINER Flyte Team <users@flyte.org>
LABEL org.opencontainers.image.source https://github.com/flyteorg/flytekit

WORKDIR /root

ARG VERSION

RUN apt-get update && apt-get install build-essential curl vim net-tools -y

COPY . /flytekit

# Pod tasks should be exposed in the default image
RUN pip install -e /flytekit
RUN pip install -e /flytekit/plugins/flytekit-deck-standard
RUN pip install -e /flytekit/plugins/flytekit-flyin

ENV PYTHONPATH "/flytekit:/flytekit/plugins/flytekit-k8s-pod:/flytekit/plugins/flytekit-deck-standard:"

RUN useradd -u 1000 flytekit
RUN chown flytekit: /root
# Some packages will create config file under /home by default, so we need to make sure it's writable
RUN chown flytekit: /home
ENV FLYTE_SDK_LOGGING_LEVEL=20
USER flytekit
```
and write a sample code
```
from flytekit import task, workflow
from flytekitplugins.flyin import vscode


@task
@vscode
def t(a: int, b: int) -> int:
    return a + b


@workflow
def wf(a: int = 5, b: int = 3) -> int:
    out = t(a=a, b=b)
    return out
```
and execute cmd:
```
docker build --no-cache --push . --build-arg PYTHON_VERSION=3.
9 -f Dockerfile.vscode -t localhost:30000/flytekit:flyin-platform
```
```
pyflyte run --remote --image localhost:30000/flytekit:flyin-pl
atform vscode_task_dockerfile.py wf --a 100 --b 50
```

### Screenshots

my machine type is `xnu-8792.81.2~2/RELEASE_ARM64_T6000 x86_64
`
so, In my local machine Flyte sanbox, the installation file that should be downloaded is of `ARM` type.
Screenshot 2023-12-03 at 7 22 36 PM

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: plugins/flytekit-flyin/flytekitplugins/flyin/vscode_lib/decorator.py [hunk 1]
```diff
@@ -112,6 +113,35 @@ def download_file(url, target_dir: Optional[str] = "."):
...
+            should be the corresponding code server information.
+
+    Returns:
+        str: The code server information corresponding to the system's architecture.
+
+    Raises:
+        ValueError: If the system's architecture is not AMD64 or ARM64.
+    """
+    machine_info = platform.machine()
+    print(f"machine_info: {machine_info}")
```
Comment (pingsutw) [inline plugins/flytekit-flyin/flytekitplugins/flyin/vscode_lib/decorator.py:136]: nit: could we use logger here instead?

**Review by pingsutw — DISMISSED** [review]

**Comment by Future-Outlier** [comment]
Test it in x86_64 machine.
image
image

**Review by Future-Outlier — DISMISSED** [review]
lgtm

**Review by Future-Outlier — APPROVED** [review]
LGTM

**Review by pingsutw — APPROVED** [review]

**Comment by troychiu** [comment]
Hi @jasonlai1218, could you add the tracking issue so that we can track all the PRs of Flyin more easily? Thank you.

**Pull request merged**